vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java

```diff
@@ -534,21 +530,58 @@ public void setDataProvider(DataProvider<T, String> dataProvider) {
             return filterConverter.apply(filterText);
         };
 
-        SerializableConsumer<C> providerFilterSlot = dataCommunicator
-                .setDataProvider(dataProvider,
-                        convertOrNull.apply(getFilterString()));
-
-        filterSlot = filter -> {
-            if (!Objects.equals(filter, lastFilter)) {
-                providerFilterSlot.accept(convertOrNull.apply(filter));
-                lastFilter = filter;
+        // Postpone data communicator provider initialization in order to
+        // trigger item count request and data fetch upon clicking on combobox
+        dataCommunicatorInitializer = () -> {
+            dataCommunicatorInitializer = null;
+            if(lazyOpenRegistration != null) {
+                lazyOpenRegistration.remove();
+                lazyOpenRegistration=null;
```

Comment:
nit: lack of spacing on assignment.
```suggestion
                lazyOpenRegistration = null;
```